dlc/src/lib.rs

```diff
@@ -0,0 +1,703 @@
+//! # Rust DLC Library
+//! Library for creating, signing and verifying transactions for the
+//! Discreet Log Contract protocol.
+//!
+
+#![crate_name = "dlc"]
+#![crate_type = "dylib"]
+#![crate_type = "rlib"]
+// Coding conventions
+#![forbid(unsafe_code)]
+#![deny(non_upper_case_globals)]
+#![deny(non_camel_case_types)]
+#![deny(non_snake_case)]
+#![deny(unused_mut)]
+#![deny(dead_code)]
+#![deny(unused_imports)]
+#![deny(missing_docs)]
+#![forbid(unsafe_code)]
+
+extern crate bitcoin;
+extern crate secp256k1;
+
+use bitcoin::blockdata::opcodes;
+use bitcoin::blockdata::script::{Builder, Script};
+use bitcoin::blockdata::transaction::{OutPoint, SigHashType, Transaction, TxIn, TxOut};
+use bitcoin::util::address::Address;
+use bitcoin::{PrivateKey, PublicKey};
+use secp256k1::ecdsa_adaptor::{AdaptorProof, AdaptorSignature};
+use secp256k1::schnorrsig::{SchnorrNonce, SchnorrSignature};
+use secp256k1::Message;
+use std::convert::TryFrom;
+
+pub mod util;
+
+const DUST_LIMIT: u64 = 1000;
+const TX_VERSION: u32 = 2;
+
+const FUND_TX_BASE_WEIGHT: usize = 286;
+const CET_BASE_WEIGHT: usize = 500;
+const P2WPKH_SPK_SIZE: usize = 22;
+const P2WPKH_WITNESS_SIZE: usize = 108;
+
+const DISABLE_LOCKTIME: u32 = 0xffffffff;
+const ENABLE_LOCKTIME: u32 = 0xfffffffe;
+
+/// Represents the payouts for a unique contract outcome. Local party represents
+/// the initiator of the contract while remote party represents the party
+/// accepting the contract.
+pub struct Outcome {
+    /// Payout for the local party
+    pub local: u64,
+    /// Payout for the remote party
+    pub remote: u64,
+}
+
+/// Contains the necessary transactions for establishing a DLC
+pub struct DlcTransactions {
+    /// The fund transaction locking both parties collaterals
+    pub fund: Transaction,
+    /// The contract execution transactions for closing the contract on a
+    /// certain outcome
+    pub cets: Vec<Transaction>,
+    /// The refund transaction for returning the collateral for each party in
+    /// case of an oracle misbehavior
+    pub refund: Transaction,
+}
+
+/// The type of an input used for funding a DLC contract. Note that it must be
+/// a segwit input type.
+pub enum InputType {
+    /// Pay to witness public key hash
+    P2WPKH,
+}
+
+impl InputType {
+    fn get_size(&self) -> usize {
+        match self {
+            InputType::P2WPKH => P2WPKH_SPK_SIZE * 4 + P2WPKH_WITNESS_SIZE,
+        }
+    }
```

Comment:
I have a question about the weight information I'm getting here.
- Why is it only returning the size of TxOut::ScriptPubKey and WitnessStack?
- Why is returning the size of TxOut::ScriptPubKey?
- It is used as PartyParams::inputs in PartyParams::get_change_output_and_fees. However, this function does not return the size of all Inputs, so it seems that only the number of TxInputs when FUND_TX_BASE_WEIGHT is calculated is assumed.
- In order to make the number of Inputs variable, it seems necessary to calculate the weight of TxInput as follows. (It is necessary to subtract the weight of TxInput from FUND_TX_BASE_WEIGHT.)
```
/// tx input base weight : (outpoint(36) + sequence(4) + scriptPubKeySize(1)) * 4
const TX_INPUT_BASE_WEIGHT: usize = 164;
fn get_weight(&self) -> usize {
  match self {
    InputType::P2WPKH => TX_INPUT_BASE_WEIGHT + P2WPKH_WITNESS_SIZE,
  }
}
```

Reply:
Thanks you are right the computation was incorrect. Updated as you suggested.